### PR TITLE
th-50: sign-up customer business validation

### PR DESCRIPTION
**Pull request by albsbz**
Created a validation schema for signing up customer business

1. Added enum with regexps
3. Added separate schemes for common user and businness, with shared rules

### CODE REVIEW

File: shared/src/libs/enums/regexp.enum.ts
Comment (Druxight) [inline shared/src/libs/enums/regexp.enum.ts:1]: we should move it to the validation folder

File: shared/src/libs/enums/user-group.enum.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const UserGroup = {
```
Comment (Druxight) [inline shared/src/libs/enums/user-group.enum.ts]: remove it

File: shared/src/packages/business/libs/types/request/business-entity.type.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+type BusinessEntity = {
+  id: number;
+  companyName: string;
+  taxNumber: string;
+  ownerId: string;
```
Comment (Druxight) [inline shared/src/packages/business/libs/types/request/business-entity.type.ts]: ownerId: number ?

File: shared/src/packages/users/libs/enums/user-validation-message.enum.ts
Comment (Druxight) [inline shared/src/packages/users/libs/enums/user-validation-message.enum.ts:1]: should be in the validation folder

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import joi from 'joi';
+
+import { UserGroup } from '~/libs/enums/user-group.enum.js';
+
+import { Regexp as REGEXP } from '../../../../libs/enums/regexp.enum.js';
```
Comment (Druxight) [inline shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: ~/libs

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    }),
+  taxNumber: joi.string().trim().required().pattern(taxNumberPattern).messages({
+    'string.empty': UserValidationMessage.REQUIRED,
+    'string.pattern.base': UserValidationMessage.TAX_NUMBER_NOT_VALID,
+  }),
+};
+
+userBusinessSignUpRules.groupId = joi
+  .number()
+  .valid(UserGroup.BUSINESS)
```
Comment (Druxight) [inline shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: Redundant validation. Leave joi.number()

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts [hunk 5]
```diff
@@ -1,21 +1,48 @@
...
+  lastName: joi.string().trim().required().pattern(namePattern).messages({
+    'string.empty': UserValidationMessage.REQUIRED,
+    'string.pattern.base': UserValidationMessage.NAME_NOT_VALID,
   }),
-  password: joi.string().trim().required(),
-  firstName: joi.string().trim().required(),
-  lastName: joi.string().trim().required(),
-  groupId: joi.number().required(),
-});
+  groupId: joi.number().valid(UserGroup.CUSTOMER).required(),
```
Comment (Druxight) [inline shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: joi.number().required()

**Review by Druxight — CHANGES_REQUESTED** [review]

File: shared/src/libs/enums/regexp.enum.ts
Comment (albsbz) [reply shared/src/libs/enums/regexp.enum.ts:1]: moved

File: shared/src/libs/enums/user-group.enum.ts (hunk 1 above)
Comment (albsbz) [reply shared/src/libs/enums/user-group.enum.ts]: done

File: shared/src/packages/business/libs/types/request/business-entity.type.ts (hunk 2 above)
Comment (albsbz) [reply shared/src/packages/business/libs/types/request/business-entity.type.ts]: sure number, and not ownerId but userId, fixed

File: shared/src/packages/users/libs/enums/user-validation-message.enum.ts
Comment (albsbz) [reply shared/src/packages/users/libs/enums/user-validation-message.enum.ts:1]: moved

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts (hunk 3 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: fixed

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts (hunk 4 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: removed groupId at all, as it not supposed to come from frontend

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts (hunk 5 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: removed groupId

File: shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+const Regexp = {
```
Comment (PavloSukhinin) [inline shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts]: Let's rename this one. UserValidationRule or something

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts [hunk 7]
```diff
@@ -1,21 +1,45 @@
...
+  }),
+  password: joi.string().trim().required().pattern(passwordPattern).messages({
+    'string.empty': UserValidationMessage.REQUIRED,
+    'string.pattern.base': UserValidationMessage.PASSWORD_NOT_VALID,
+  }),
   email: joi
     .string()
     .trim()
+    .min(5)
+    .max(254)
```
Comment (PavloSukhinin) [inline shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: Let's combine all user validation constants in one enum

File: frontend/package.json [hunk 8]
```diff
@@ -18,6 +18,7 @@
     "@fortawesome/fontawesome-svg-core": "6.4.2",
     "@fortawesome/free-solid-svg-icons": "6.4.2",
     "@fortawesome/react-fontawesome": "0.2.0",
+    "@googlemaps/react-wrapper": "1.1.35",
```
Comment (kalinkaaaa14) [inline frontend/package.json]: Merge latest development changes please

File: shared/src/packages/users/libs/validation-schemas/validation-schemas.ts [hunk 9]
```diff
@@ -1 +1,2 @@
+export { userBusinessSignUp } from './user-business-sign-up.validation-schema.js';
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/validation-schemas.ts]:
```suggestion
export { businessSignUp } from './business-sign-up.validation-schema.js';
```

File: shared/src/packages/users/libs/validation-schemas/validation-schemas.ts [hunk 10]
```diff
@@ -1 +1,2 @@
...
 export { userSignUp } from './user-sign-up.validation-schema.js';
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/validation-schemas.ts]: customerSignUp sounds better to me

File: shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+const UserValidationMessage = {
+  REQUIRED: 'This field is mandatory',
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts]:
```suggestion
  FIElD_IS_REQUIRED: 'This field is mandatory',
```

File: shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import { type BusinessEntityT } from '~/packages/business/business.js';
+
+import { type UserSignUpRequestDto } from './user-sign-up-request-dto.type.js';
+
+type UserBusinessSignUpRequestDto = UserSignUpRequestDto &
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts]:
```suggestion
type BusinessSignUpRequestDto = CustomerSignUpRequestDto &
```

File: shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+import { type BusinessEntityT } from '~/packages/business/business.js';
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts]: Why BusinessEntityT is exported, why not BusinessEntity 🤔

File: shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts (hunk 6 above)
Comment (kalinkaaaa14) [reply shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts]: Agree

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import joi from 'joi';
+
+import { type UserBusinessSignUpRequestDto } from '../types/types.js';
+import { Regexp as REGEXP } from './enums/regexp.enum.js';
+import { UserValidationMessage } from './enums/user-validation-message.enum.js';
+import { userSignUpRules } from './user-sign-up.validation-schema.js';
+
+const companyNamePattern = new RegExp(REGEXP.COMPANY_NAME);
+const taxNumberPattern = new RegExp(REGEXP.TAX_NUMBER);
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: I don't like that for each pattern you are writing a separate variable in this file. Maybe it's better to create some helper for wrapping to new RegExp(expression)

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts [hunk 15]
```diff
@@ -1,21 +1,45 @@
 import joi from 'joi';
 
-import { UserValidationMessage } from '../enums/enums.js';
 import { type UserSignUpRequestDto } from '../types/types.js';
+import { Regexp as REGEXP } from './enums/regexp.enum.js';
+import { UserValidationMessage } from './enums/user-validation-message.enum.js';
 
-const userSignUp = joi.object<UserSignUpRequestDto, true>({
+const passwordPattern = new RegExp(REGEXP.PASSWORD);
+const namePattern = new RegExp(REGEXP.NAME);
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: Same for these patterns

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts [hunk 16]
```diff
@@ -1,21 +1,45 @@
...
+  lastName: joi.string().trim().required().pattern(namePattern).messages({
+    'string.empty': UserValidationMessage.REQUIRED,
+    'string.pattern.base': UserValidationMessage.NAME_NOT_VALID,
+  }),
+};
+
+const userSignUp = joi.object<UserSignUpRequestDto, true>(userSignUpRules);
 
-export { userSignUp };
+export { userSignUp, userSignUpRules };
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: Do you need to export signUp rules?

Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: I need it as rules shared between signup customer and business, as an option I can move them to separate file

File: shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts (hunk 13 above)
Comment (albsbz) [reply shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts]: Used at the same maner as we use UserEntityT, may be to visualy destinguish type from class

File: shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts (hunk 14 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-business-sign-up.validation-schema.ts]: something like `createRegexep=(regexp)=>new RegExp(regexp)`? Or I can use `password: joi.string().trim().required().pattern(new RegExp(Rule.PASSWORD)).messages({
    'string.empty': Message.REQUIRED,
    'string.pattern.base': Message.PASSWORD_NOT_VALID,
  }),`  I added variables at top of the file just for code readability. I'm not sure what approach is better

File: shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts (hunk 6 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/enums/regexp.enum.ts]: renamed to UserValidationRules

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts (hunk 7 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: moved to user-validation-rules.enum

File: frontend/package.json (hunk 8 above)
Comment (albsbz) [reply frontend/package.json]: done

File: shared/src/packages/users/libs/validation-schemas/validation-schemas.ts (hunk 9 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/validation-schemas.ts]: done

File: shared/src/packages/users/libs/validation-schemas/validation-schemas.ts (hunk 10 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/validation-schemas.ts]: renamed

File: shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts (hunk 11 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts]: renamed

File: shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts (hunk 12 above)
Comment (albsbz) [reply shared/src/packages/users/libs/types/user-business-sign-up-request-dto.type.ts]: renamed

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts (hunk 15 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: commented above

File: shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts (hunk 16 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/user-sign-up.validation-schema.ts]: moved common rules to separate file

File: shared/src/packages/users/libs/validation-schemas/enums/user-validation-rules.enum.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+const UserValidationRules = {
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/enums/user-validation-rules.enum.ts]:
```suggestion
const UserValidationRule = {
```

Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/enums/user-validation-rules.enum.ts]: fixed

File: shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+const UserValidationMessage = {
+  FIELD_IS_REQUIRED: 'This field is mandatory',
+  EMAIL_NOT_VALID: 'This email does not seem valid',
+  PASSWORD_NOT_VALID: 'Must be 6+ characters, at least 1 letter and 1 number',
+  NAME_NOT_VALID: 'Must be 1 - 40 characters, start with a letter, no digits',
+  PHONE_NOT_VALID: 'Must be 8-17 digits, start with +',
```
Comment (PavloSukhinin) [inline shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts]:
```suggestion
  NAME_NOT_VALID: 'Must be 1 - 40 characters, start with a letter, no digits',
  PHONE_NOT_VALID: 'Must be 8 - 17 digits, start with +',
```
just to be consistent

**Review by PavloSukhinin — CHANGES_REQUESTED** [review]

File: shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts (hunk 18 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/enums/user-validation-message.enum.ts]: fixed

File: shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+  UserValidationMessage as Message,
+  UserValidationRule as Rule,
+} from '../enums/enums.js';
+
+const passwordPattern = new RegExp(Rule.PASSWORD);
+const namePattern = new RegExp(Rule.NAME);
+const phonePattern = new RegExp(Rule.PHONE);
+
+const signUpRules = {
+  phone: joi.string().trim().required().pattern(phonePattern).messages({
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts]:
```suggestion
  phone: joi.string().trim().required().pattern(createPatternByRule(rule)).messages({
```
 It's better to add some helper which will return new RegExp(rule), so we can avoid to write each time each variable for new pattern in this file

File: shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    'string.empty': Message.FIELD_IS_REQUIRED,
+    'string.pattern.base': Message.PHONE_NOT_VALID,
+  }),
+  password: joi.string().trim().required().pattern(passwordPattern).messages({
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts]: Reuse helper for patterns here

File: shared/src/packages/users/libs/validation-schemas/business-sign-up.validation-schema.ts [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+const companyNamePattern = new RegExp(Rule.COMPANY_NAME);
+const taxNumberPattern = new RegExp(Rule.TAX_NUMBER);
+
+const userBusinessSignUpRules = {
+  ...commonSignUpRules,
+  companyName: joi
+    .string()
+    .trim()
+    .required()
+    .pattern(companyNamePattern)
```
Comment (kalinkaaaa14) [inline shared/src/packages/users/libs/validation-schemas/business-sign-up.validation-schema.ts]: Same about new pattern

File: shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts (hunk 19 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts]: done

File: shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts (hunk 20 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/common-rules/signup-rules.ts]: done

File: shared/src/packages/users/libs/validation-schemas/business-sign-up.validation-schema.ts (hunk 21 above)
Comment (albsbz) [reply shared/src/packages/users/libs/validation-schemas/business-sign-up.validation-schema.ts]: done

**Review by PavloSukhinin — APPROVED** [review]

**Review by Druxight — APPROVED** [review]

**Pull request merged**